bootstrap/cmd/bootstrap/app/sourceRepos.go

```diff
@@ -0,0 +1,236 @@
+package app
+
+import (
+	"fmt"
+	"github.com/cenkalti/backoff"
+	"io/ioutil"
+	//"github.com/otiai10/copy"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/sourcerepo/v1"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"os"
+	"path"
+	"time"
+)
+
+// SourceRepo manages a local clone of a source repo.
+type SourceRepo struct {
+	// The GCP project
+	project string
+
+	// Local path for the repo
+	localDir string
+
+	repoName string
+
+	ts oauth2.TokenSource
+	r  *git.Repository
+}
+
+const (
+	RemoteName = "gcpsourcerepo"
+)
+
+// NewSourceRepo initializes a repo object for the specified local directory.
+// If the directory doesn't exist an error is raised.
+//
+//
+// If the remote repo doesn't exist it is created.
+//
+// TODO(jlewi): What if the repo exists but the directory doesn't? In that case should
```

Comment:
Current behavior is `repo/version/deploy_name/actual_configs`.
So all deployments on v0.6 will go to same source repo.

We can always check if repo exists, and clone if it does.
Then we might need to create sub folder based on deploy name; update configs; and commit & push whole repo.

Reply:
Why not make it one repo per deployment? Doesn't that simplify things at least to start with?

We can always add support later for putting multiple deployments into the same repository.

Reply:
I think NewSourceRepo() is compatible with having multiple apps stored in the same repo.
i.e. we could still have
repo/version/deploy_name/actual_configs

To do that we would just set the app_dir to

app_dir = path.join(local_dir, version, deploy_name)

I'm leaning towards making it 1 repo per app at least to start with. The main reason is I think that simplifies things like dealing with merge conflicts and force pushes. If we have multiple apps per repo I think there is more likely to be multiple writers to the repo which makes it harder to reason about.

Also, I'd like to do all the source repo clones and pushes out of band with the actual deployment so that we don't block deployment based on cloning and pushing.

I believe we can do this even if we want to store multiple apps in the same repo but I think its harder to reason about.

Reply:
What if the repo already exists? Might be leftover from kubeflow older version, or even user happens to create one by themselves.

Reply:
Depending on how we use NewSourceRepo in KfctlServer it would be overwritten.

Specifically here's how this would currently work

1. KfctlServer creates  ${KFAPP}
1. NewSourceRepo  runs `git init` inside ${KFAPP}
1. NewSourceRepo creates SourceRepo if it doesn't exist
1. KfctlServer will do a force push to ${KFAPP}/master 

So whatever is in the repo would be overwritten.

Reply:
I assume we'll keep repo for user's custom change after they delete the cluster?
Currently `number of repo` quota per project is 3.
We'll likely hit quota issue if we create new repo for every deployment.

Reply:
Good point. We should definitely reuse repos in that case. I don't think that changes any of the code in this PR.

I think it will just change how we integrate it into kfctl sever.

I think each app should push to its own branch so that we are less likely to get conflicting changes but we can resolve details when we integrate into kfctl server.